astropy/io/fits/fitsrec.py

```diff
@@ -92,9 +92,10 @@ def __setitem__(self, key, value):
             if indx < self.start or indx > self.end - 1:
                 raise KeyError(f"Key '{key}' does not exist.")
         elif isinstance(key, slice):
-            for indx in range(slice.start, slice.stop, slice.step):
-                indx = self._get_indx(indx)
-                self.array.field(indx)[self.row] = value
+            for indx, val in zip(range(*key.indices(self.array._nfields)), value):
+                indx = self._get_index(indx)
+                self.array.field(indx)[self.row] = val
```

Comment:
Looks to me like two different variables competing over a single name, which we can avoid by simply not rebinding
```suggestion
                self.array.field(self._get_index(indx))[self.row] = val
```

Reply:
I renamed the variable to avoid making the line too complex.